src/fsharp/LowerStateMachines.fs

```diff
@@ -0,0 +1,895 @@
+     phase2: (Map<int, ILCodeLabel> -> Expr)
+
+     /// The labels allocated for this portion of the computation
+     entryPoints: int list
+
+     /// The state variables allocated for one portion of the sequence expression (i.e. the local let-bound variables which become state variables)
+     stateVars: ValRef list
+
+     /// All this values get represented via the 'this' pointer
+     thisVars: ValRef list
+
+     /// The vars captured by the non-synchronous resumable path
+     asyncVars: FreeVars
+   }
+
+let sm_verbose = try System.Environment.GetEnvironmentVariable("FSharp_StateMachineVerbose") <> null with _ -> false
+
+let rec (|OptionalResumeAtExpr|) g expr =
+    match expr with
+    | IfUseResumableStateMachinesExpr g (OptionalResumeAtExpr g res, _) -> res
+    | Expr.Sequential(ResumeAtExpr g pcExpr, codeExpr, NormalSeq, _, _m) -> (Some pcExpr, codeExpr)
+    | _ -> (None, expr)
+
+/// Implement a decision to represent a 'let' binding as a non-escaping local variable (rather than a state machine variable)
+let RepresentBindingAsTopLevelOrLocal (bind: Binding) (res2: StateMachineConversionFirstPhaseResult) m =
+    if sm_verbose then 
+        printfn "LowerStateMachine: found local variable %s" bind.Var.DisplayName
+
+    { res2 with
+        phase1 = mkLetBind m bind res2.phase1 
+        phase2 = (fun ctxt -> mkLetBind m bind (res2.phase2 ctxt)) }
+
+/// Implement a decision to represent a 'let' binding as the 'this' pointer of the state machine,
+/// because it is rebinding the 'this' variable
+let RepresentBindingAsThis (bind: Binding) (res2: StateMachineConversionFirstPhaseResult) _m =
+    if sm_verbose then 
+        printfn "LowerStateMachine: found local variable %s" bind.Var.DisplayName
+
+    { res2 with
+        thisVars = mkLocalValRef bind.Var :: res2.thisVars
+        // Drop the let binding on the floor as it is only rebinding the 'this' variable
+        phase1 = res2.phase1 
+        phase2 = res2.phase2 }
+
+/// Implement a decision to represent a 'let' binding as a state machine variable
+let RepresentBindingAsStateVar (bind: Binding) (res2: StateMachineConversionFirstPhaseResult) m =
+    if sm_verbose then 
+        printfn "LowerStateMachine: found state variable %s" bind.Var.DisplayName
+    
+    let (TBind(v, e, sp)) = bind
+    let sp, spm =
+        match sp with
+        | DebugPointAtBinding.Yes m -> DebugPointAtSequential.Both, m
+        | _ -> DebugPointAtSequential.StmtOnly, e.Range
+    let vref = mkLocalValRef v
+    { res2 with
+        phase1 = mkSequential sp m (mkValSet spm vref e) res2.phase1
+        phase2 = (fun ctxt ->
+            let generate2 = res2.phase2 ctxt
+            let generate =
+                //mkCompGenSequential m
+                    (mkSequential sp m
+                        (mkValSet spm vref e)
+                        generate2)
+                    // TODO: zero out the current value to free up its memory - but return type is not unit...
```

Comment:
Thanks, well spotted, I'm testing completing this

Reply:
I've solved this in recent commits and added tests for codegen

Reply:
I'll have to revert the addition of this clearing out of fields of the state machine for now.

The problem is, it should only be done on the synchronous return paths (result of resumable code = true).  However when converting the state machine code we can't actually be sure we even have such a return value, nor that this is its interpretation.  

I'm not quite sure what to do about this.  The sequence expression generation has sufficient knowledge to do this - for example it emits jumps rather than returning `true` return values.  I'm not sure we do, it would be good to be in a position to do this in the future.    We can perhaps determine when we are expanding the ResumableCode type (= return value bool in resumable code) and check return values and emit sets for those cases.

Reply:
> We can perhaps determine when we are expanding the ResumableCode type (= return value bool in resumable code) and check return values and emit sets for those cases.

I'm trialling a solution to this along these lines

Reply:
This is now fixed.  I've added codegen tests